pages/k8s/upgrading.md

```diff
@@ -0,0 +1,319 @@
+|stable	        | The latest stable released version of Kubernetes                 |
+|candidate  |	Release candidate versions of Kubernetes                              |
+|beta	           | Latest alpha/beta of Kubernetes for the specified release|
+|edge	          | Nightly builds of the specified release of Kubernetes          |
+
+For most use cases, it is strongly recommended to use the 'stable' version of charms.
+
+### Upgrading the **kube-api-loadbalancer**
+
+A core part of **CDK** is the kubeapi-load-balancer component. A mismatch of versions
+could have an effect on API availability and access controls. To ensure API service
+continuity this upgrade should precede any upgrades to the **Kubernetes** master
+and worker units.
+
+```bash
+juju upgrade-charm kubeapi-load-balancer
+```
+
+### Upgrading the **kubernetes-master** units
+
+To start upgrading the Kubernetes master units, first upgrade the charm:
+
+```bash
+juju upgrade-charm kubernetes-master
+```
+
+Once the charm has been upgraded, it can be configured to select the desired
+**Kubernetes** channel, which takes the form `Major.Minor/risk-level`.  This is
+then passed as a configuration option to the charm. So, for example, to select
+the stable 1.10 version of **Kubernetes**, you would enter:
+
+```bash
+juju config kubernetes-master channel=1.10/stable
+```
+
+If you wanted to try a release candidate for 1.12, the channel would be `1.12/candidate`.
+
+<div class="p-notification--caution">
+  <p class="p-notification__response">
+    <span class="p-notification__status">Note:</span>
+Once the configuration has been changed, the charms will be put into a `blocked` state.
+You must continue the upgrade process, even if you revert the configuration to the
+currently active version of Kubernetes.
+  </p>
+</div>
+
+Once the desired version has been configured, the upgrades should be performed.
+This is done by running the `upgrade` action on each master unit in the
+cluster:
+
+```bash
+juju run-action kubernetes-master/0 upgrade
+juju run-action kubernetes-master/1 upgrade
+```
+
+If you have more master units in your cluster, you should continue and run this
+process on every one of them.
+
+You can check the progress of the upgrade by running:
+
+```bash
+juju status | grep master
+```
+
+Ensure that all the master units have upgraded and are reporting normal status
+before continuing to upgrade the worker units.
+
+### Upgrading the **kubernetes-worker** units
+
+For a running cluster, there are two different ways to proceed:
+
+ * [Blue-green][blue-green] upgrade - This requires more resources, but should ensure a
+    safe, zero-downtime transition of workloads to an updated cluster
+ * In-place upgrade - this simply upgrades the workers in situ, which may involve some
+    service interruption but doesn't require extra resources
+
+ Both methods are outlined below. The blue-green method is recommended for
+ production systems.
+
+#### Blue-green upgrade
+
```

Comment:
The upstream instructions for this are garbage - you can tell just by skimming them that there are half a dozen problems with the outlined method there, which leads me to think that nobody ever actually tried it.
I have tried to implement this method in a more logical way. It _seemed_ to work when I ran through it, but I wasn't using a realistic workload on the cluster. Happy to accept any input or suggestions on this section if someone has actually **done** this. I will attempt to check with a more realistic setup at some point.

Reply:
Excellent, thank you.